// @peculiar/asn1-cms/build/cjs/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignatureAlgorithmIdentifier = exports.DigestAlgorithmIdentifier = exports.CMSVersion = void 0;
const tslib_1 = require("tslib");
const asn1_x509_1 = require("@peculiar/asn1-x509");
const asn1_schema_1 = require("@peculiar/asn1-schema");
var CMSVersion;
(function (CMSVersion) {
    CMSVersion[CMSVersion["v0"] = 0] = "v0";
    CMSVersion[CMSVersion["v1"] = 1] = "v1";
    CMSVersion[CMSVersion["v2"] = 2] = "v2";
    CMSVersion[CMSVersion["v3"] = 3] = "v3";
    CMSVersion[CMSVersion["v4"] = 4] = "v4";
    CMSVersion[CMSVersion["v5"] = 5] = "v5";
})(CMSVersion = exports.CMSVersion || (exports.CMSVersion = {}));
;
let DigestAlgorithmIdentifier = class DigestAlgorithmIdentifier extends asn1_x509_1.AlgorithmIdentifier {
};
DigestAlgorithmIdentifier = tslib_1.__decorate([
    asn1_schema_1.AsnType({ type: asn1_schema_1.AsnTypeTypes.Sequence })
], DigestAlgorithmIdentifier);
exports.DigestAlgorithmIdentifier = DigestAlgorithmIdentifier;
let SignatureAlgorithmIdentifier = class SignatureAlgorithmIdentifier extends asn1_x509_1.AlgorithmIdentifier {
};
SignatureAlgorithmIdentifier = tslib_1.__decorate([
    asn1_schema_1.AsnType({ type: asn1_schema_1.AsnTypeTypes.Sequence })
], SignatureAlgorithmIdentifier);
exports.SignatureAlgorithmIdentifier = SignatureAlgorithmIdentifier;
